shared/message/input/JoystickEvent.proto

```diff
@@ -0,0 +1,87 @@
+/*
+ * This file is part of the NUbots Codebase.
+ *
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2013 NUbots <nubots@nubots.net>
+ */
+
+syntax = "proto3";
+
+package message.input;
+
+import "google/protobuf/timestamp.proto";
+
+message JoystickEvent {
+    enum EventType {
+        UNKNOWN = 0;
+        BUTTON  = 1;    // Button press/release
+        AXIS    = 2;    // Joystick movement
+        INIT    = 128;  // Initial state of device
+    }
+
+    enum Axis {
+        LEFT_JOYSTICK_HORIZONTAL  = 0;
+        LEFT_JOYSTICK_VERTICAL    = 1;
+        L2                        = 2;
+        RIGHT_JOYSTICK_HORIZONTAL = 3;
+        RIGHT_JOYSTICK_VERTICAL   = 4;
+        R2                        = 5;
+
+        // ACCEL_Y = 23;
+        // ACCEL_X = 24;
+        // ACCEL_Z = 25;
+    }
+
+    enum Button {
+        CROSS          = 0;
+        CIRCLE         = 1;
+        TRIANGLE       = 2;
+        SQUARE         = 3;
+        L1             = 4;
+        R1             = 5;
+        SELECT         = 8;
+        START          = 9;
+        PS             = 10;
+        LEFT_JOYSTICK  = 11;
+        RIGHT_JOYSTICK = 12;
+        DPAD_UP        = 13;
+        DPAD_DOWN      = 14;
+        DPAD_LEFT      = 15;
+        DPAD_RIGHT     = 16;
+    }
+
+    /**
+     * The timestamp of the event
+     */
+    google.protobuf.Timestamp timestamp = 1;
+
+    /**
+     * The value associated with this joystick event.
+     * For buttons this will be either 1 (down) or 0 (up).
+     * For axes, this will range between -32768 and 32767.
+     */
+    int32 value = 2;
+
+    /**
+     * The event type.
+     */
+    EventType type = 3;
+
+    /**
+     * The axis/button number.
+     */
+    Axis   axis   = 4;
```

Comment:
The axis event and button event should probably be separate messages. Maybe a separate message per button type/axis type.